### ISSUE TITLE
Feature request: some pre-set options for SQL style guides

### ISSUE THREAD

**Issue by mrshu** [sqlfluff/sqlfluff#3]
Hi @alanmcruickshank,

Thanks a ton again for taking the time to build `sqlfluff` -- I believe it has the potential to be the [linter that is sorely missing](https://www.reddit.com/r/SQL/comments/8wnysv/is_it_accurate_to_say_that_uniform_code_style_and/)!

I am not sure what are your future plans, but if I may, I'd suggest trying to add rules from a few guidelines -- here are the ones I am familiar with:

- http://docs-origin.telemetry.mozilla.org/concepts/sql_style.html
- https://www.sqlstyle.guide/

If you would be open to exploring this direction, I would be happy to put together a few rules as well -- from looking at the code it seems fairly straightforward.

Thanks again!

**Comment by alanmcruickshank** [comment]
Hi @mrshu  - totally up for this. I'm about to embark on rebuilding some of the parsing engine to deal with some of these types of rules. Definitely keen to move in this direction.

What I want to have is a modular approach where we have  few default style guide, but people can also plug and play their own style guides as they wish.

**Comment by mrshu** [comment]
Awesome, thanks a ton @alanmcruickshank!

Let me know if I can be of any help.

**Comment by alanmcruickshank** [comment]
@mrshu - I'm going to update this to a feature request around having pre-set configuration settings to allow users to choose from a selection of pre-existing style guides rather than always having to go with the default.

Setting expectations I don't think this will be a feature we get to very soon - the initial style guide is still pretty bare bones so not much to choose between - but I would like to support this eventually.

**Comment by mrshu** [comment]
Sure thing @alanmcruickshank, that makes a lot of sense -- thanks!

**Comment by nolanbconaway** [comment]
deserves noting that the DBT style guide is one worth implementing once the project is at this stage!

https://github.com/fishtown-analytics/corp/blob/master/dbt_coding_conventions.md

**Comment by alanmcruickshank** [comment]
@mrshu @nolanbconaway 

I'm moving this issue to the 0.3.0 mileston rather than 0.2.0 (which should be coming out soon). I realised before we start supporting *multiple* potential styles, I need to get sqlfluff to support *one* consistent style first.

The way this will eventually be implemented will be by having some pre-set config files which can be optionally loaded, by some kind of `style` argument - that then loads the relevant config file (packaged as part of the `sqlfluff` package), and it overrides some config values in the default config file - and therefore overrides the default style.

**Comment by mrshu** [comment]
Thanks @alanmcruickshank! 

Feel free let me know if there are parts of it you could see outside help as being of some use -- I'd be happy to provide some.

**Comment by alanmcruickshank** [comment]
@mrshu @nolanbconaway [and anyone else for that matter]

Could both of you provide some examples or your internal style guides?

I'm thinking there are four approaches I'd like to go with here:
1. The sqlfluff defaults. There should be only one set of these, and so far I've tried to tread a fairly mainstream path, taking a lot of inspiration from the dbt style guide (among others). This isn't very useful if your org style differs from the defaults however!
2. [potentially] some inbuilt (batteries included style) alternate styles which would be selectable. These would be maintained as part of the core sqlfluff, and ideally modelled on a public sql spec. These could be organisation specific, but I don't want to maintain loads of these, and so I think I would only inlcude them if they were a) popular and b) different enough to the defaults to warrant it.
3. Installable/pluggable style guides maintained *outside* the core sqlfluff library. In #164 there's a discussion of a plugin layer, and I would be very up for any organisation publishing their own style guide as a sqlfluff compatible plugin which members of their team can then _install_ and use.
4. Organisation/project config files. This already exists, where a project can have it's own `.sqlfluff` file to specify the configurations to use within that project.

Are there any other sensible options I've missed?

If people roughly agree, what I'm trying to work out is:
- Are there are style guides which warrant being in the second category? (Given the assumption that the defaults will be very similar to the dbt style guide, albeit perhaps a little more lenient).
- Would people otherwise feel comfortable with option 3 or option 4? If so which would they be more likely to use?

**Comment by pwildenhain** [comment]
@alanmcruickshank I think that the `.sqlfluff` config file should meet any needs for organization specific style. At my work, I implemented `sqlfluff` in CI in part of our codebase. During this we first reconsidered our style when it wasn't line with the `.sqlfluff` defaults. Then, if we decided to deviate, simply customized the `.sqlfluff` to match our organization specific style.

If people have a style they want to support, which is not configurable, then they can file an issue/PR similar to #240 

This also has the benefit of being already done 😁 and if it turns out that it's really not working, then we could always pivot later on

**Comment by alanmcruickshank** [comment]
@pwildenhain - I like that approach. For someone coming to to the tool do you think that the documentation on how to actually do this is clear enough? If so I might close this issue for now, otherwise we should make sure that the guidance is at least in the right place and then close it.

**Comment by pwildenhain** [comment]
I think it's pretty clear in the ["Getting Started"](https://docs.sqlfluff.com/en/latest/gettingstarted.html#custom-usage) section that people can customize according their organization's needs. 

I submitted a PR which adds a link to the default configuration section which should be the cherry on top 🍒 🍨

**Comment by mrshu** [comment]
@alanmcruickshank To me a `.sqlfluff` is all that's necessary, at least for the usecase I had in mind when I first opened this issue (nearly two years ago no less!).

**Comment by mrshu** [comment]
@alanmcruickshank To me a `.sqlfluff` is all that's necessary, at least for the usecase I had in mind when I first opened this issue (nearly two years ago no less!).
